### PR TITLE
Pointer arithmetic

### LINKED ISSUE #230: Implement high-level pointer arithmetic and relational operations
At the moment high-level dialect does not permits mixed arithmetic with integers and pointer types.
Similar applies for relational operations.

Preliminary work done on: https://github.com/trailofbits/vast/tree/pointer-binary-ops

We want to allow arithmetic operations to take mix of integer and pointer types. This requires to fix constraints,
ideally not just to `AnyType`. 

This also applies to family of assignment binary operations. For them we also want to check whether they preserve correct return type (ideally infer it from parameters).

Questions to investigate:

- is multiplication/division/remainder allowed for pointer types?
- are unsigned predicates right way to compare pointers?

### PR DESCRIPTION
**Pull request by Jezurko**
Fix #230

### CODE REVIEW

File: include/vast/Dialect/HighLevel/HighLevelOps.td [hunk 1]
```diff
@@ -620,8 +620,42 @@ def ImplicitCastOp   : CastOp< "implicit_cast", [] >;
 def CStyleCastOp     : CastOp< "cstyle_cast", [] >;
 def BuiltinBitCastOp : CastOp< "builtin_bitcast", [] >;
 
+class IsInteger< string arg > : PredOpTrait< "value is of an integer type",
+    CPred< "$" # arg # ".getType().isa< hl::IntType>()" >
```
Comment (xlauko) [inline include/vast/Dialect/HighLevel/HighLevelOps.td:624]: fix spacing in between `< ... >` on multiple places in this file

File: include/vast/Dialect/HighLevel/HighLevelOps.td [hunk 2]
```diff
@@ -620,8 +620,42 @@ def ImplicitCastOp   : CastOp< "implicit_cast", [] >;
...
+>;
+
+class IsLong< string arg > : PredOpTrait< "value is of an integer type",
```
Comment (xlauko) [inline include/vast/Dialect/HighLevel/HighLevelOps.td:627]: copy-paste of description

File: include/vast/Dialect/HighLevel/HighLevelOps.td [hunk 3]
```diff
@@ -620,8 +620,42 @@ def ImplicitCastOp   : CastOp< "implicit_cast", [] >;
...
+
+class IsPointerAdditive< string lhs, string rhs, string res > : PredOpTrait< "IsAdditive",
+    And< [Or< [And< [IsPointer< lhs >.predicate, IsInteger< rhs>.predicate] >,
+               And< [IsInteger< lhs >.predicate, IsPointer< rhs >.predicate] >] >,
+         IsPointer< res >.predicate
+    ] >
+>;
+
+class IsPointerSub< string lhs, string rhs, string res > : PredOpTrait< "subtracting two pointers",
+    And< [IsPointer< lhs >.predicate, IsPointer< rhs >.predicate, IsLong< res >.predicate] >
```
Comment (xlauko) [inline include/vast/Dialect/HighLevel/HighLevelOps.td:643]: maybe `PtrDiffLike`?

File: include/vast/Dialect/HighLevel/HighLevelOps.td [hunk 4]
```diff
@@ -633,28 +667,50 @@ class ArithBinOp< string mnemonic, list< Trait > traits = [] >
...
+    let skipDefaultBuilders = 1;
+    let builders = [ OpBuilder<(ins "Value":$lhs, "Value":$rhs), [{
+        $_state.addOperands(lhs);
+        $_state.addOperands(rhs);
+        auto type = lhs.getType();
+        if(rhs.getType().isa< hl::PointerType >())
+        {
+            // Both operands are pointers -> it's a subtraction
+            if(type.isa< hl::PointerType >())
+                type = odsBuilder.getType< hl::LongType >();
```
Comment (xlauko) [inline include/vast/Dialect/HighLevel/HighLevelOps.td:682]: this is not correct we need to go through our type builder

I suggest to take result type in constructor and transform it before in visitor

File: include/vast/Dialect/HighLevel/HighLevelOps.td [hunk 5]
```diff
@@ -633,28 +667,50 @@ class ArithBinOp< string mnemonic, list< Trait > traits = [] >
...
-def DivUOp : ArithBinOp<"udiv", []>;
-def DivFOp : ArithBinOp<"fdiv", []>;
-def RemSOp : ArithBinOp<"srem", []>;
-def RemUOp : ArithBinOp<"urem", []>;
-def RemFOp : ArithBinOp<"frem", []>;
+class StandardArithBinOp< string mnemonic, list< Trait > traits = [] >
+    : ArithBinOp< mnemonic, !listconcat(traits, [SameOperandsAndResultType]) >;
+
+
+def AddIOp : ArithBinOp< "add", [Commutative, IsAdditive< "lhs", "rhs", "result" >]>;
```
Comment (xlauko) [inline include/vast/Dialect/HighLevel/HighLevelOps.td:697]: can not you have addition of mixed types?

Comment (xlauko) [reply include/vast/Dialect/HighLevel/HighLevelOps.td:697]: similarly subtraction?

File: include/vast/Dialect/HighLevel/HighLevelOps.td [hunk 6]
```diff
@@ -722,6 +778,22 @@ class ShiftOp< string mnemonic, list< Trait > traits = [] >
...
+class IsCompoundPtrAdditive< string lhs, string rhs, string res > : PredOpTrait< "IsAdditive",
+    And< [IsLValuePointer< lhs >.predicate,
+          IsInteger< rhs>.predicate,
+          IsPointer< res >.predicate ] >
+>;
+
+class IsCompoundAdditive< string lhs, string rhs, string res > : PredOpTrait< "is additive operation",
+    Or<[
+        IsCompoundPtrAdditive< lhs, rhs, res >.predicate,
+        CPred< "$" # lhs # ".getType().cast< LValueType >().getElementType() ==$" # res # ".getType()" >
```
Comment (xlauko) [inline include/vast/Dialect/HighLevel/HighLevelOps.td:794]: space after ==

**Review by xlauko — COMMENTED** [review]
add test on mixed operations of integer types

```
int + long
short + long
ptr + short
...
```

File: include/vast/Dialect/HighLevel/HighLevelOps.td (hunk 5 above)
Comment (Jezurko) [reply include/vast/Dialect/HighLevel/HighLevelOps.td:697]: I think that in that case the right hand side should be casted to match the types. At least that's what happens everywhere else. Why should it be different here?

**Review by xlauko — APPROVED** [review]

**Pull request merged**